File: 2026-07/src/day_28/prms_172927/Solution2.java
package day_28.prms_172927;

import java.util.*;

class Solution2 {
    public int solution(int[] picks, String[] minerals) {
        int pickCnt = picks[0] + picks[1] + picks[2];
        int[][] cnt = new int[pickCnt][3];

        for (int i = 0; i < Math.min(minerals.length, pickCnt * 5); i++) {
            String mineral = minerals[i];

            if (mineral.equals("diamond")) {
                cnt[i / 5][0]++;
            } else if (mineral.equals("iron")) {
                cnt[i / 5][1]++;
            } else {
                cnt[i / 5][2]++;
            }
        }

        Arrays.sort(cnt, (o1, o2) -> {
            if (o1[0] != o2[0]) return Integer.compare(o2[0], o1[0]);
            return Integer.compare(o2[1], o1[1]);
        });

        int ans = 0;
        for (int[] arr : cnt) {
            if (picks[0] > 0) {
                ans += arr[0] + arr[1] + arr[2];
                picks[0]--;
            } else if (picks[1] > 0) {
                ans += arr[0] * 5 + arr[1] + arr[2];
                picks[1]--;
            } else if (picks[2] > 0) {
                ans += arr[0] * 25 + arr[1] * 5 + arr[2];
                picks[2]--;
            } else {
                break;
            }
        }

        return ans;
    }
}

File: 2026-07/src/day_28/prms_172927/Solution2.cpp
#include <algorithm>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

int solution(vector<int> picks, vector<string> minerals) {
    int pickCnt = picks[0] + picks[1] + picks[2];
    vector<tuple<int, int, int>> v(pickCnt);

    for (int i = 0; i < min((int)minerals.size(), pickCnt * 5); i++) {
        string mineral = minerals[i];
        auto& [di, ir, st] = v[i / 5];

        if (mineral == "diamond") {
            di++;
        } else if (mineral == "iron") {
            ir++;
        } else {
            st++;
        }
    }

    sort(v.rbegin(), v.rend());

    int ans = 0;
    for (auto [di, ir, st] : v) {
        if (picks[0] > 0) {
            ans += di + ir + st;
            picks[0]--;
        } else if (picks[1] > 0) {
            ans += di * 5 + ir + st;
            picks[1]--;
        } else if (picks[2] > 0) {
            ans += di * 25 + ir * 5 + st;
            picks[2]--;
        } else {
            break;
        }
    }

    return ans;
}
